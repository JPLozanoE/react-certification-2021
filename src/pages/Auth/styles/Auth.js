import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const Title = styled.h3`
  margin-bottom: 20px;
`;

export const Main = styled.div`
  align-items: center;
  display: flex;
  justify-content: center;
  margin: 0px;
  height: 94vh;
  max-height: 100%;
  width: 100%;
`;

export const BoxContainer = styled.div`
  background-color: white;
  border: 1px solid gray;
  box-shadow: 0px 3px #cccccc;
  border-radius: 2px;
  padding: 20px;
  width: 250px;
`;

export const StyledLink = styled(Link)`
  text-decoration: none;
  color: #6b6b6b;

  &:hover {
    text-decoration: underline;
  }
`;
