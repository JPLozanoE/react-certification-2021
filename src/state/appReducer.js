import { types } from '../types/types';

export const appReducer = (state = {}, action) => {
  switch (action?.type) {
    case types.changeSearchQuery:
      return {
        ...state,
        search: action.payload,
      };

    case types.setSelectedVideo:
      return {
        ...state,
        selectedVideo: action.payload,
      };

    case types.toggleTheme:
      return {
        ...state,
        isDarkTheme: !state.isDarkTheme,
      };

    case types.setRecommendedVideos:
      return {
        ...state,
        videos: action.payload,
      };

    case types.login:
      return {
        ...state,
        auth: {
          isLogged: true,
          displayName: action.payload.displayName,
          uid: action.payload.uid,
        },
      };

    case types.logout:
      return {
        ...state,
        auth: {
          isLogged: false,
          displayName: '',
          uid: null,
        },
      };

    case types.addFavoriteVideo:
      return {
        ...state,
        favoriteVideos: [
          ...state.favoriteVideos,
          { ...action.payload, isFavorite: true },
        ],
      };

    case types.deleteFavoriteVideo:
      return {
        ...state,
        videos: state.videos.map((video) =>
          video.id.videoId === action.payload ? { ...video, isFavorite: false } : video
        ),
        favoriteVideos: state.favoriteVideos.filter(
          (favoriteVideo) => favoriteVideo.id.videoId !== action.payload
        ),
      };

    case types.updateRecommendedVideos:
      return {
        ...state,
        videos: state.videos.map((video) =>
          video.id.videoId === action.payload ? { ...video, isFavorite: true } : video
        ),
      };

    default:
      return state;
  }
};
