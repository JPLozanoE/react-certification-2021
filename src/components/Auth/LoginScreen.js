/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable react/jsx-filename-extension */
import React, { useContext, useState } from 'react';
import validator from 'validator';
import { firebase, googleAuthProvider } from '../../firebase/firebase-config';
import { useForm } from '../../hooks/useForm';
import { StyledLink } from '../../pages/Auth/styles/Auth';
import { Button, Input, SocialNetworksDiv, Title, GoogleDiv } from './styles/AuthStyles';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';

export const LoginScreen = () => {
  const { dispatch } = useContext(AppContext);
  const [msgError, setMsgError] = useState('');

  const [formValues, handleInputChange] = useForm({
    email: '',
    password: '',
  });

  const { email, password } = formValues;
  const isFormValid = () => {
    if (!validator.isEmail(email)) {
      setMsgError('Email is not valid');
      return false;
    }

    if (password.length < 6) {
      setMsgError('Invalid password');
      return false;
    }

    setMsgError('');
    return true;
  };

  const handleLogin = (e) => {
    e.preventDefault();
    isFormValid();
    console.log(email, password);
    if (isFormValid()) {
      firebase
        .auth()
        .signInWithEmailAndPassword(email, password)
        .then(({ user }) => {
          console.log(user);
          dispatch({
            type: types.login,
            payload: { displayName: user.displayName, uid: user.uid },
          });
        })
        .catch((error) => {
          console.error(error.message);
        });
    }
  };

  const handleGoogleLogin = () => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch({
          type: types.login,
          payload: { displayName: user.displayName, uid: user.uid },
        });
        console.log(user);
      });
  };

  return (
    <>
      <Title>Login</Title>
      <form onSubmit={handleLogin}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <Input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <Input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />

        <Button
          // {loading && (disabled={true})}
          disabled={false}
          className="btn btn-primary btn-block mb-1"
          type="submit"
        >
          Login
        </Button>

        {/* <hr/> */}
        <SocialNetworksDiv className="auth__social-networks mt-1">
          <p>Login with Social Networks</p>
          <GoogleDiv className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </GoogleDiv>
        </SocialNetworksDiv>
        <StyledLink to="/auth/register" className="link">
          Create new account
        </StyledLink>
      </form>
    </>
  );
};
