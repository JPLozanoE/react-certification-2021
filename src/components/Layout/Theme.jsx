import React, { useContext } from 'react';
import { ThemeProvider } from 'styled-components';
import { AppContext } from '../../state/AppContext';

const Theme = ({ children }) => {
  const {
    state: { isDarkTheme },
  } = useContext(AppContext);
  const theme = {
    main: {
      ...(isDarkTheme ? { backgroundColor: '#181818' } : { backgroundColor: '#ffffff' }),
    },
    titles: {
      ...(isDarkTheme ? { color: '#ffffff' } : { color: '#00000' }),
    },
    subtitles: {
      ...(isDarkTheme ? { color: 'rgb(170,170,170)' } : { color: 'rgb(96,96,96)' }),
    },
  };

  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export default Theme;
