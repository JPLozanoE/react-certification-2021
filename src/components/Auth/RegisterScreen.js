/* eslint-disable react/jsx-filename-extension */
import React, { useContext, useState } from 'react';
import validator from 'validator';
import { firebase } from '../../firebase/firebase-config';
import { useForm } from '../../hooks/useForm';
import { StyledLink } from '../../pages/Auth/styles/Auth';
import { Input, Title } from './styles/AuthStyles';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import { Button } from '../../__globalStyles';

export const RegisterScreen = () => {
  const { dispatch } = useContext(AppContext);
  const [msgError, msgSetError] = useState('');

  const [{ name, email, password, password2 }, handleInputChange] = useForm({
    name: '',
    email: '',
    password: '',
    password2: '',
  });
  const isFormValid = () => {
    let error = false;

    if (name.trim().length === 0) {
      console.log('Name is required');
      error = true;
      msgSetError('Name is required');
      return false;
    }

    if (!validator.isEmail(email)) {
      console.log('Email is not valid');
      error = true;
      msgSetError('Email is not valid');

      return false;
    }

    if (password !== password2 || password.length < 5) {
      console.log('Password should be at least 6 characters and match eachother');
      error = true;
      msgSetError('Password should be at least 6 characters and match eachother');

      return false;
    }
    msgSetError('');
    return error === false;
  };

  const handleRegister = (e) => {
    e.preventDefault();
    if (isFormValid()) {
      firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
        .then(async ({ user }) => {
          await user.updateProfile({ displayName: name });
          console.log(user);
          dispatch({
            type: types.login,
            payload: { displayName: user.displayName, uid: user.uid },
          });
        })
        .catch((error) => {
          console.error(error.message);
        });
      console.log('Formulario correcto');
    }
  };

  return (
    <>
      <Title className="auth__title">Register</Title>
      <form onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}

        <Input
          className="auth__input"
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <Input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <Input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />

        <Input
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          name="password2"
          value={password2}
          onChange={handleInputChange}
        />

        <Button className="btn btn-primary btn-block mb-5" type="submit">
          Register
        </Button>
        <div>
          <StyledLink to="/auth/login" className="link">
            Already registered?
          </StyledLink>
        </div>
      </form>
    </>
  );
};
