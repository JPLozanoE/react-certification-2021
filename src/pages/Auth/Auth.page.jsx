import React from 'react';
import { BoxContainer, Main } from './styles/Auth';

export const AuthPage = ({ children }) => {
  return (
    <Main>
      <BoxContainer>{children}</BoxContainer>
    </Main>
  );
};
