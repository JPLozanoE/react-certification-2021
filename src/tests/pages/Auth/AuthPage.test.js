/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { AppContext } from '../../../state/AppContext';
import Theme from '../../../components/Layout/Theme';
import { demoState } from '../../fixtures/demoState';
import { AuthPage } from '../../../pages/Auth/Auth.page';
import { LoginScreen } from '../../../components/Auth/LoginScreen';

describe('Auth Page', () => {
  let wrapper;

  test('Should render properly', () => {
    const providerValues = {
      state: { ...demoState },
      dispatch: jest.fn(),
    };
    wrapper = render(
      <AppContext.Provider value={providerValues}>
        <Theme>
          <Router>
            <AuthPage>
              <LoginScreen />
            </AuthPage>
          </Router>
        </Theme>
      </AppContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.debug();
    expect(wrapper.getByRole('heading', { name: 'Login' })).toBeInTheDocument();
  });
});
