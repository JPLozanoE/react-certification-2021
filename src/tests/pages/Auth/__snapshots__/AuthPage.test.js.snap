// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Page Should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gtsrHT kUvrCA"
      >
        <div
          class="sc-dlnjwi eEaXHH"
        >
          <h3
            class="sc-eCApnc hGciCd"
          >
            Login
          </h3>
          <form>
            
            <input
              autocomplete="off"
              class="sc-jSFjdj bBOMsQ auth__input"
              name="email"
              placeholder="Email"
              type="text"
              value=""
            />
            <input
              class="sc-jSFjdj bBOMsQ auth__input"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
            <button
              class="sc-pNWdM kPTJIB btn btn-primary btn-block mb-1"
              type="submit"
            >
              Login
            </button>
            <div
              class="sc-gKAaRy gDVUFA auth__social-networks mt-1"
            >
              <p>
                Login with Social Networks
              </p>
              <div
                class="sc-iCoGMd xsXRv google-btn"
              >
                <div
                  class="google-icon-wrapper"
                >
                  <img
                    alt="google button"
                    class="google-icon"
                    src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                  />
                </div>
                <p
                  class="btn-text"
                >
                  <b>
                    Sign in with google
                  </b>
                </p>
              </div>
            </div>
            <a
              class="sc-hKFxyN cFwZJz link"
              href="/auth/register"
            >
              Create new account
            </a>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gtsrHT kUvrCA"
    >
      <div
        class="sc-dlnjwi eEaXHH"
      >
        <h3
          class="sc-eCApnc hGciCd"
        >
          Login
        </h3>
        <form>
          
          <input
            autocomplete="off"
            class="sc-jSFjdj bBOMsQ auth__input"
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
          <input
            class="sc-jSFjdj bBOMsQ auth__input"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            class="sc-pNWdM kPTJIB btn btn-primary btn-block mb-1"
            type="submit"
          >
            Login
          </button>
          <div
            class="sc-gKAaRy gDVUFA auth__social-networks mt-1"
          >
            <p>
              Login with Social Networks
            </p>
            <div
              class="sc-iCoGMd xsXRv google-btn"
            >
              <div
                class="google-icon-wrapper"
              >
                <img
                  alt="google button"
                  class="google-icon"
                  src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                />
              </div>
              <p
                class="btn-text"
              >
                <b>
                  Sign in with google
                </b>
              </p>
            </div>
          </div>
          <a
            class="sc-hKFxyN cFwZJz link"
            href="/auth/register"
          >
            Create new account
          </a>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
